changefeedccl: add roachtest testing msk serverless

asg0451 opened the pull request:
Add a roachtest to test the kafka sink against msk serverless.

Resolves: #124431

Release note: None

Commit:
workload: add sending stats to histogram

Sends histogram latency to a localhost listener over udp to a custom
port. The other side can process the latency however it wants.

Epic: CRDB-37514

Release note: None

nvb committed:
syncutil: make IntMap generic

Closes #103147.

This commit makes syncutil.IntMap generic, so that it can be used with
keys other than int64. This restores some of the generality lost when
the standard library's sync.Map was specialized in our fork to IntMap in
8340120c. It does so without losing the performance benefits that
specialization provided, because we use generics instead of interface
objects. This limited use of generics with no virtual function calls
performs identically to the hand-specialized version.

The use of generics also provides type safety. Callers no longer need to
cast int-like keys to and from int64. They also no longer need to cast
typed-pointers to and from unsafe.Pointer.

Benchmark performance is meaningfully unchanged:
```
name                                        old time/op    new time/op    delta
LoadMostlyHits/*syncutil.IntMap-10            2.05ns ± 6%    2.33ns ± 4%  +14.01%  (p=0.000 n=10+10)
LoadMostlyMisses/*syncutil.IntMap-10          1.82ns ± 5%    2.00ns ± 7%   +9.88%  (p=0.000 n=10+9)
LoadOrStoreBalanced/*syncutil.IntMap-10        206ns ± 1%     208ns ± 2%     ~     (p=0.059 n=8+8)
LoadOrStoreUnique/*syncutil.IntMap-10          394ns ± 3%     410ns ±10%     ~     (p=0.203 n=9+10)
LoadOrStoreCollision/*syncutil.IntMap-10      1.33ns ± 5%    1.54ns ± 4%  +16.20%  (p=0.000 n=10+10)
LoadAndDeleteBalanced/*syncutil.IntMap-10     2.41ns ± 4%    2.66ns ± 2%  +10.14%  (p=0.000 n=10+10)
LoadAndDeleteUnique/*syncutil.IntMap-10       0.65ns ±41%    0.70ns ±44%     ~     (p=0.143 n=10+10)
LoadAndDeleteCollision/*syncutil.IntMap-10    0.84ns ± 3%    0.78ns ±44%     ~     (p=0.481 n=10+10)
Range/*syncutil.IntMap-10                      982ns ± 5%    1022ns ± 1%   +4.05%  (p=0.001 n=10+10)
AdversarialAlloc/*syncutil.IntMap-10           158ns ± 0%     157ns ± 1%     ~     (p=0.722 n=7+10)
AdversarialDelete/*syncutil.IntMap-10         30.6ns ±10%    29.2ns ± 5%   -4.60%  (p=0.004 n=10+9)
DeleteCollision/*syncutil.IntMap-10           0.65ns ±56%    0.88ns ±47%  +35.02%  (p=0.007 n=10+10)

name                                        old alloc/op   new alloc/op   delta
LoadMostlyHits/*syncutil.IntMap-10             0.00B          0.00B          ~     (all equal)
LoadMostlyMisses/*syncutil.IntMap-10           0.00B          0.00B          ~     (all equal)
LoadOrStoreBalanced/*syncutil.IntMap-10        49.3B ± 8%     49.2B ± 8%     ~     (p=0.854 n=10+9)
LoadOrStoreUnique/*syncutil.IntMap-10           101B ± 5%       67B ± 6%  -33.64%  (p=0.000 n=8+9)
LoadOrStoreCollision/*syncutil.IntMap-10       0.00B          0.00B          ~     (all equal)
LoadAndDeleteBalanced/*syncutil.IntMap-10      0.00B          0.00B          ~     (all equal)
LoadAndDeleteUnique/*syncutil.IntMap-10        0.00B          0.00B          ~     (all equal)
LoadAndDeleteCollision/*syncutil.IntMap-10     0.00B          0.00B          ~     (all equal)
Range/*syncutil.IntMap-10                      0.00B          0.00B          ~     (all equal)
AdversarialAlloc/*syncutil.IntMap-10           27.0B ± 0%     27.0B ± 0%     ~     (all equal)
AdversarialDelete/*syncutil.IntMap-10          8.40B ±19%     8.00B ± 0%     ~     (p=0.211 n=10+10)
DeleteCollision/*syncutil.IntMap-10            0.00B          0.00B          ~     (all equal)

name                                        old allocs/op  new allocs/op  delta
LoadMostlyHits/*syncutil.IntMap-10              0.00           0.00          ~     (all equal)
LoadMostlyMisses/*syncutil.IntMap-10            0.00           0.00          ~     (all equal)
LoadOrStoreBalanced/*syncutil.IntMap-10         0.00           0.00          ~     (all equal)
LoadOrStoreUnique/*syncutil.IntMap-10           1.00 ± 0%      1.00 ± 0%     ~     (all equal)
LoadOrStoreCollision/*syncutil.IntMap-10        0.00           0.00          ~     (all equal)
LoadAndDeleteBalanced/*syncutil.IntMap-10       0.00           0.00          ~     (all equal)
LoadAndDeleteUnique/*syncutil.IntMap-10         0.00           0.00          ~     (all equal)
LoadAndDeleteCollision/*syncutil.IntMap-10      0.00           0.00          ~     (all equal)
Range/*syncutil.IntMap-10                       0.00           0.00          ~     (all equal)
AdversarialAlloc/*syncutil.IntMap-10            0.00           0.00          ~     (all equal)
AdversarialDelete/*syncutil.IntMap-10           0.00           0.00          ~     (all equal)
DeleteCollision/*syncutil.IntMap-10             0.00           0.00          ~     (all equal)
```

Release note: None

nvb committed:
syncutil: rename int_mapXXX.go files to mapXXX.go

File renames.

Epic: None
Release note: None

nvb committed:
syncutil: rename IntMap to Map

Type rename.

Epic: None
Release note: None

rafiss committed:
cli: add --shutdown flag to `node drain`

Release note (cli change): The `cockroach node drain` command now has an
optional --shutdown flag. When set, the node will shutdown after
draining successfully completes.

annrpom committed:
sql: ensure truncate preserves pkey zone configs

Previously, we did not copy zone configs over
from old indexes to new ones created during a
truncate for pkeys. This was due to truncate
re-using the same logic as our `PrimaryKeySwap`
mutation. This patch adds an override in said logic to
allow for zcs to be copied from old to new indexes.

Epic: CRDB-22729
Fixes: #94188

Release note (bug fix): Fixed bug where zone configs
on pkeys disappeared during truncate.

Commit:
roachtest: add perturbation tests

This commit adds a set of roachtests and a framework to measure the
impact to latency of various perturbations to the system.

The roachtest names all have the format
perturbation/[full|dev|metamorphic]/[perturbationType]

The four perturbationType variations added in this initial commit are:
restart
partition
add-node
decommission

Epic: CRDB-37514

Release note: None

Commit:
workload: split reads from follower reads

Follower reads and non-follower reads have different performance
characteristics and should be treated differently.

Release note: None

Epic: CRDB-37514

Commit:
roachtest: add download to http util

This commit adds a method Download to RoachtestHTTPClient. This allows
easily saving a file from a crdb admin UI by setting up the necessary
auth headers.

Epic: CRDB-37514

Release note: None

Commit:
roachtest: add utility methods for gathering profiles

This commit adds two new utility methods for gathering profiles. The
first, profileTopStatements will begin profiling common statements that
exceed a given latency. It is intended to be run near the beginning of a
test after the workload has started to start gathering profiles from
common statements. The second downloadProfiles will download statements
at the end of the test and put them in the artifacts direction.

Epic: CRDB-37514

Release note: None

Commit:
workload: add histogram latency protobuf

Add a protobuf used for sending a single latency measurement to a remote
observer.

Release note: None

Epic: CRDB-37514

Commit:
roachtest: add utility method to get local ip

This commit adds a method to get the IP address of the node running
roachtest. This is useful for streaming information back to that node
from the workload nodes.

Epic: CRDB-37514

Release note: None

shubhamdhama committed:
distsql: improve unhealthy node detection in SQL instances-based planning.

The distsql planner currently uses different strategies to avoid planning
on unhealthy nodes in single and multi-tenant deployments. For
single-tenant deployments, we check for a healthy gRPC connection and use
information from the gossip protocol, which allows for quick detection. For
multi-tenant deployments, we rely on the sql_instances system table cache
to check if an instance is healthy. However, this cache may not update for
over half a minute (40 seconds in my tests) after a node goes down, leading
to planning on an unhealthy instance and retrying the corresponding
subquery locally on the gateway.

To prevent planning on dead SQL instances, the resolver now ensures it does
not consider such instances by first establishing a healthy gRPC
connection. This means that in shared-process mode, if an instance is found
to be unhealthy, the gateway will be used. In external mode, such a node
would be excluded from the candidates for locality-based matching or
round-robin selection.

Fixes: #120774

Epic CRDB-39091

Release note: None

pav-kv committed:
raft: add safety checks in truncateAndAppend

This commit adds safety checks in unstable.truncateAndAppend which
ensure the last accepted term does not regress, and the last index in
the log regresses only if we are accepting a higher-term write. This
maintains an invariant necessary for the log write protocol, and guards
the log from accepting writes that are raft-incorrect.

Epic: none
Release note: none

pav-kv committed:
raft: make unstable a logSlice

This commit converts the (offset, entries) pair in the unstable
structure into a type-safe logSlice. One immediate benefit of doing so
is that unstable now knows the lastEntryID at all times, whereas
previously it would fall back to fetching it from Storage interface when
empty.

The second benefit is introducing the logSlice.term field into the
unstable's posession. This field carries strong semantics: the log is
consistent with the leader log at this term. This allows implementing
safety checks in truncateAndAppend method guarding, preventing log
truncations when not expected. Log truncations in raft can only happen
when accepting a higher-term log append.

Epic: none
Release note: none

yuzefovich committed:
sql: improve bundle with plan-gist matching

Previously, if the bundle collection was triggered via the plan-gist
matching, we would include "-- plan elided due to gist matching" string
as `plan.txt` since we couldn't set up the optimizer for the complete
bundle collection. However, this is suboptimal since it requires the
user to copy the gist elsewhere. This commit improves the bundle
collection a bit by explicitly including the gist into `plan.txt` as
well as decoding it (on a best effort basis) in to the file.

Release note: None

annrpom committed:
scpb: modify TableZoneConfig element

This commit modifies the TableZoneConfig element
that will become relevant for supporting
`ALTER TABLE ... CONFIGURE ZONE` in the DSC.

Informs: #117574
Release note: None

annrpom committed:
schemachanger: add table zone config opgen rule

This patch adds an opgen rule for tbl zone configs
in the ADD direction.

Informs: #117574
Release note: None

annrpom committed:
schemachanger: support tbl zone config

This patch adds the functionality to configure
a zone configuration on a table.

Fixes: #126495
Informs: #117574
Release note: None

separating for review

TODO squash into previous commit

navsetlur committed:
logical: Add replanning logic to LDR

Currently there's no replanning logic for logical replication, which means
there is no handling for changes to the cluster (e.g. nodes being added or failing).
This change uses the PhysicalPlanChecker, which at regular intervals compares the
initial plan with a more up to date plan and throws an error whne they differ too much,
leading to a replan for the actual logical replication.

Release note: none
Part of: https://cockroachlabs.atlassian.net/browse/CRDB-39122

yuzefovich committed:
scripts: fix a couple of nits in ldr

Release note: None

nvb committed:
kv: consolidate range lease validation

Closes #123498.
Informs #118435.

This commit extracts lease request and lease transfer validation into
the `kv/kvserver/leases` package which was introduced in #124682. This
avoids the scattered leases validity checks which were previously spread
between `replica_range_lease.go`, `replica_proposal_buf.go`
`cmd_lease_request.go`, `cmd_lease_transfer.go`.

This makes the logic easier to understand and easier to test. It will
also make the logic easier to adjust when we make the following two
changes in service of #118435:
- default to rejecting lease requests on unknown leaders
- perform lease request validation above raft, like we do for lease
  transfers

This second change also ties in with Leader leases (#123847). Until we
make that change, we will occasionally need to propose expiration-based
leases (e.g. when we're not the leader) just for them to be rejected by
the proposal buffer.

Release note: None

navsetlur committed:
logical: Add ReplanCount to logical metrics

annrpom committed:
sql: override StmtTimeout to 0 for background jobs

For background jobs and schema changes, our Internal
Executor inherits from cluster settings. If a statement
timeout is set cluster-wide, then background jobs will
respect this -- which does not make much sense; thus,
this patch overrides the StmtTimeout to 0s for background
jobs.

Fixes: #126261
Release note (bug fix): Background jobs no longer respect
a statement timeout.

annrpom committed:
scop: remove unused SeqNum from AddDatabaseZoneConfig

The `SeqNum` is not needed during immediate mutation.

Release note: None

xinhaoz committed:
ui: fix query fetching store ids for db/tables

These queries were updated in #118904 to make them more performant.
In that change we mistakenly removed the null check before reading
the `rows` field, causing pages using this request to crash if this
query fails.

Epic: none
Fixes: #126348

Release note (bug fix): Database overview and db details
pages should not crash if the range information is not
available.

annrpom committed:
scbuild: add ResolvePhysicalTable

This patch adds `ResolvePhysicalTable` to
the name resolver interface. This will allow
us to set zone configurations on materialized
views and sequences -- which are both tables
that need to be stored in the kv layer.

Informs: #117574

Release note: None

kyle-a-wong committed:
log,job: log status change on update via updater

Adds a new `status_change` job event that logs
on every status update.

Epic: CRDB-37556

Release note (general change): Job status changes
now log events to the OPS channel, indicating the
previous status and the new status.

navsetlur committed:
crosscluster: Refactor measurePlanChange to be reusable

This code can now be used by both physical and logical replication
to measure distSQL plan divergence.

Epic: none
Release note: none

pav-kv committed:
raft: take accTerm from unstable

The last accepted term is now tracked in the unstable logSlice.term
field. This commit removes the redundant raft.accTerm field, and uses
the unstable.term instead.

Epic: none
Release note: none

pav-kv committed:
raft: propagate logSlice to truncateAndAppend

Epic: none
Release note: none

pav-kv committed:
raft: unify table-driven test pattern in log_unstable_test.go

Epic: none
Release note: none

asg0451 committed:
changefeedccl: implement new Kafka sink to sidestep reordering issue

Add a new Kafka sink utilizing the franz-go
library and our batchingSink wrapper that does not
suffer from the row reordering issue that the
existing one does. It is behind a cluster setting
(`changefeed.new_kafka_sink_enabled`) which is
disabled by default. We intend to backport this
change. To that end, we use the previous version
of franz-go (v1.6.1) to not require bumping
important dependencies.

Epic: CRDB-38892
Resolves: #122372

Release note (enterprise change): Add a new Kafka
sink behind a cluster setting
(`changefeed.new_kafka_sink_enabled`, disabled by
default).

msbutler committed:
streamingccl: skip TestStreamingAutoReplan under deadlock

Fixes #125991

Release note: none

asg0451 committed:
changefeedccl: fix sink leaks in a few places

There are some places where we neglect to close
sinks, which lead to goroutine leaks and thus test
failures in the new Kafka sink PR.

Epic: CRDB-38892

Release note: none

mgartner committed:
sql: rename PreparedStatement.Memo to BaseMemo

Release note: None

mgartner committed:
sql: refactor plan_opt.go

The logic for picking a prepared memo has been moved to a function,
`fetchPreparedMemoLegacy`. This makes it easy in the following commit
to fallback to this logic based on the `plan_cache_mode` setting.

Release note: None

mgartner committed:
sql: implement plan_cache_mode=force_generic_plan

The `force_generic_plan` option for the session setting
`plan_cache_mode` has been implemented. Under this setting, prepared
statements will reuse a fully optimized, generic query plan, if a
non-stale one exists. If such a plan does not exist or is stale, a new
generic query plan will be built.

The term "generic query plan" is used in user-facing contexts to
represent any plan that is fully optimized without knowledge of specific
placeholder values nor folding stable expressions. Therefore, all of the
following are considered "generic":

  1. A plan for a query with no placeholders nor fold-able stable
     expressions.
  2. A plan utilizing the placeholder fast-path.
  3. A plan fully optimized without replacing placeholders with values.

However, only (3) is currently stored in the
`PreparedStatement.GenericMemo` field. This was a deliberate decision
that allows falling back to the original prepared statement logic when
switching between `force_custom_plan` and `force_generic_plan`. This
will make backports with this commit less risky. My goal, in a future
commit, is to put all generic query plans in the `GenericMemo` field to
reduce the confusion between the user-facing definition of "generic" and
the internal logic.

This commit does not change the behavior of the query cache. Fully
optimized plans for statements without placeholders or placeholder
fast-path plans are still cached in the query cache. Generic query plans
optimized in the presence of placeholders are not cached.

Release note (sql change): The session setting
`plan_cache_mode=force_generic_plan` can now be used to force prepared
statements to use query plans that are optimized once and reused in
future executions without re-optimization, as long as it does not become
stale due to schema changes or a collection of new table statistics. The
setting takes effect during `EXECUTE` commands. `EXPLAIN ANALYZE`
includes a "plan type" field. If a generic query plan is optimized for
the current execution, the "plan type" will be "generic, re-optimized".
If a generic query plan is reused for the current execution without
performing optimization, the "plan type" will be "generic, reused".
Otherwise, the "plan type" will be "custom".

pav-kv committed:
raft: use logSlice in log_unstable_test.go

This commit is a mechanical refactor in log_unstable_test.go. It adds a
newUnstableForTesting helper, and converts all the tests in the file to
use logSlice for initializing the unstable structure.

Epic: none
Release note: none

pav-kv committed:
raft: straighten up maybeAppend and append semantics

Epic: none
Release note: none

pav-kv committed:
raft: add unstable.append with stricter semantics

Epic: none
Release note: none

nvb committed:
kv: add support for Lease.MinExpiration

Closes #125235.

This commit begins using the new `Lease.MinExpiration` field, which was
introduced in 7781cfa6. `Lease.MinExpiration` is assigned to epoch-based
leases (and soon, leader leases) during upgrades from expiration-based
lease. It is a simple, direct way to ensure that the expiration of the
lease does not regress during such promotions.

We previously solved this issue in 6dd54b46 by manually heartbeating the
node liveness record when we detected an expiration regression. This was
backwards compatible, so it was a better backport target, but it was
also complex, as demonstrated by the need for b2ac52a4. In the future,
we will be able to remove the manual heartbeat logic, as it is no longer
necessary once MinTimestamp is supported by all clusters.

Epic: None

aadityasondhi committed:
metric: add prev, next values in counter update panic

This should help investigate potential bugs in metrics.

Informs #126470.

Release note: None

rail committed:
build: update go to 1.22.5
* [x] Update `build/teamcity/internal/release/build-and-publish-patched-go/impl.sh` with the new version and adjust SHA256 sums as necessary.
* [x] Adjust `GO_VERSION` and `GO_FIPS_COMMIT` for the FIPS Go toolchain ([source](./teamcity/internal/release/build-and-publish-patched-go/impl-fips.sh)).
* [x] Run the `Internal / Cockroach / Build / Toolchains / Publish Patched Go for Mac` build configuration in TeamCity with your latest version of the script above. Note the job depends on another job `Build and Publish Patched Go`. That job prints out the SHA256 of all tarballs, which you will need to copy-paste into `WORKSPACE` (see below). `Publish Patched Go for Mac` is an extra step that publishes the *signed* `go` binaries for macOS. That job also prints out the SHA256 of the Mac tarballs in particular.
* [x] Adjust `--@io_bazel_rules_go//go/toolchain:sdk_version` in [.bazelrc](../.bazelrc).
* [x] Bump the version in `WORKSPACE` under `go_download_sdk`. You may need to bump [rules_go](https://github.com/bazelbuild/rules_go/releases). Also edit the filenames listed in `sdks` and update all the hashes to match what you built in the step above.
* [x] Bump the version in `WORKSPACE` under `go_download_sdk` for the FIPS version of Go (`go_sdk_fips`).
* [x] Run `./dev generate bazel` to refresh `distdir_files.bzl`, then `bazel fetch @distdir//:archives` to ensure you've updated all hashes to the correct value.
* [x] Bump the go version in `go.mod`.
* [x] Bump the default installed version of Go in `bootstrap-debian.sh` ([source](./bootstrap/bootstrap-debian.sh)).
* [x] Replace other mentions of the older version of go (grep for `golang:<old_version>` and `go<old_version>`).

Epic: none
Release note: None

arulajmani committed:
cli: bump default worker pool to heavy

For the benefit of TestDebugCheckStore.

Closes #122328
Closes #126593

Epic: none

Release note: None

kyle-a-wong committed:
log: add new parameter to log.StructuredEvent

adds a severity parameter to
`log.StructuredEvent` to give more fine
grained control of logged structured events

Epic: none
Release note: None

Nukitt committed:
build: add script to build fuzzers in oss-fuzz

nicktrav committed:
github: update codeowners

Update all instances of `repl-prs` to `kv-prs`.

Release note: None.

mgartner committed:
sql: fix pg_catalog formatting of BYTES expressions

This commit fixes a regression introduced in #126297 that caused
incorrect formatting of BYTES expressions in pg_catalog.

Fixes #126344

There is no release note because this bug is not present in any
releases.

Release note: None

mgartner committed:
sqlsmith: unskip TestGenerateParse

Release note: None

nameisbhaskar committed:
roachtest: add manual compaction operation

This change is to add a new operation that will enable us to run
 manual compaction. Having a manual compaction ongoing would induce
 a moderate amount of chaos, and could possibly lead to interesting
 behaviour in Storage as well as elsewhere in CockroachDB.

 Fixes: #126480
 Epic: None

fqazi committed:
catalog/lease: address TestDescriptorRefreshOnRetry flake

Previously, it was possible for this test to flake if the stats
collection ran after table creation, which would cause the release count to
not match the acquire count. To address this, this patch will disable
automatic stats collection for this leasing test.

Fixes: #125976

Release note: None

lyang24 committed:
kv-obs: annotate consistency queue graph with explanation

This commit annotates the consistency queue graph with a hint that explains
what the role of the queue is and that a persistent pending backlog is expected
and benign.

Fixes: #https://github.com/cockroachdb/cockroach/issues/83869

Release note: None

Commit:
multitenant: allow auto-upgrades to internal versions

Previously, tenants would only auto upgrade to non-internal
versions (`Internal` field is zero). The rationale is that we only
want tenants to start attempting to upgrade when the storage cluster
has finished its own upgrades.

However, allowing auto upgrades to internal versions is more
consistent with how the storage cluster performs auto upgrades. It's
also very helpful while testing upgrades on master.

In this commit, we allow tenants to auto upgrade to any version, as
long as we know that the storage cluster version is the same as the
latest cluster version known in the tenant's binary.

Fixes: #115045

Release note: None

Commit:
roachtest: clean up acceptance tests

Acceptance tests are tests that are run as part of CI. As such, they
need to be compatible with roachtest local, otherwise there's no point
in declaring them as "acceptance".

This commit enforces that requirement, and removes tests that are
registered as `acceptance` but don't fit these criteria.

Epic: none

Release note: None

Commit:
roachtest: use `CompatibleClouds` to define clouds where a test can run

This removes the vast majority of cloud checks during test run
time. These checks typically verify whether the test is running in
specific (set of) clouds, and then skip the test if it is not.

However, that is purpose of the `CompatibleClouds` field. Declaring
cloud compatibility in the cloud spec has the added benefit that the
test runner won't assign a cluster to run the test, only to have the
test skip itself shortly after.

Epic: none

Release note: None

pav-kv committed:
raft: rm "offset" from log_unstable_test.go

Epic: none
Release note: none

pav-kv committed:
raft: add safety checks to snapshot restore

This commit adds safety checks to the snapshot restore stack, ensuring
that we don't accidentally lose durability of entries that are truncated
on applying a snapshot.

Epic: none
Release note: none

dt committed:
cli: fix duplicate scheme in encode-uri

Previously 'encode-uri postgres://some.url' would return a URI with the
scheme duplicated, since it assumed it was not passed a scheme, however
it is common that it is.

Release note: none.
Epic: none.

dt committed:
cli: add --user option to encode-uri

Release note: none.
Epic: none.

dt committed:
cli: add --certs-dir support to encode-uri

Release note (ops change): the encode-uri command now supports the --certs-dir option as an alternative to passing individual cert paths.

Epic: none.

dt committed:
crosscluster/logical: add metrics for reason event sent to DLQ

Release note: none.
Epic: none.

dt committed:
crosscluster/logical: fix retry queue guage drift on unclean proc exit

Release note: none.
Epic: none.

rohitsanj committed:
Release note (general change): Fixed typo in array_functions description

dt committed:
crosscluster/logical: avoid age limit for unclosed levels

Release note: none.
Epic: none.

sambhav-jain-16 committed:
roachprod-microbench: add clean command

Epic: none

Release note: None

Previously cleaning benchmark output was only internally exposed in roachprod-microbench, but we now need it as part of a CI job, hence a new utility command "clean" has been added to roachprod-microbench
See: https://github.com/cockroachdb/cockroach/issues/106661

dt committed:
crosscluster/logical: add cluster settings for retry queue

Release note: none.
Epic: none.

sumeerbhola committed:
storage: make pebbleLogger.Eventf also log

This is useful when tracing is not enabled, but verbosity of logging can
be increased. Also, we currently have gaps in context propagation in
tracing which this can help mitigate for scenarios where slowness is
reproducible.

Relates to https://github.com/cockroachdb/pebble/issues/3728

Epic: none

Release note: None

Uzair5162 committed:
sql: ignore outer buckets when merging partial and full statistics

We merge partial stats with full stats by prepending partial
histogram buckets with UpperBound less than the first full histogram
bucket and appending partial histogram buckets with UpperBounds greater
than the last full histogram bucket. This doesn't work when outer
buckets exist in either histogram since they have max or min valued
UpperBounds, which overlap with the other statistic's buckets.

This commit fixes this by removing outer buckets from both histograms
before merging them, and then adjusting counts at the end to add outer
buckets back if necessary.

Fixes: #126822

See also: #125950

Release note (bug fix): Fixed a bug when creating partial statistics
using extremes (which is disabled by default) where it would
occasionally fail to merge the new partial stat with existing full
stats. This occurs when outer buckets were added to either
statistic to account for extra distinct count.

sumeerbhola committed:
storage: remove unnecessarily strict RangeBounds assertion

pebble.Iterator permits calling `RangeBounds` when there is no range key.

Epic: none

Release note: None

aadityasondhi committed:
kvserver: add mutex to prevent store metrics race

Previously, `store.computeMetrics()` was being called by concurrent
goroutines that would collect and update engine metrics. The update
could happen in reverse order of the collection and would lead to
non-monotonic counters.

Fixes #126470.

Release note: None

Uzair5162 committed:
sql: fix and improve the computation of merged statistic counts

Previously, we were iterating over the buckets of the merged
histogram to compute its RowCount and DistinctCount. We were casting
each bucket's DistinctRange count from a float to an int in this loop,
which resulted in a final distinct count that was less than expected for
the merged stat for larger tables.

This commit fixes this and simplifies the computation of the merged stat
counts by using the partial and full stat counts instead of iterating
over the merged histogram.

Fixes: #126730

See also: #125950

Release note (bug fix): Fixed a bug when creating partial statistics
using extremes (which is disabled by default) where it would
result in a merged statistic with inaccurate distinct counts. Also made
the merged statistic computation more efficient.

yuzefovich committed:
sql: add pgvector in a couple of missing places

This commit expands the recently added support for PGVector to the
following places:
- PG binary encoding is now supported (see
https://github.com/pgvector/pgvector/blob/8772c8de68cf11fba59b381db8f262951222358f/src/vector.c#L413-L430
for the source code)
- `json_build_object` builtin.

It also adds a comment for why we consciously didn't add support for
PGVector in CDC avro format.

Release note: None

rafiss committed:
tree: avoid race on Overload.PreferredOverload

We need to modify the PreferredOverload field if there is a UDF with the
same name as a builtin function. However, we should make a copy so that
we avoid modifying the hardcoded builtin overload definition.

Release note: None

cockroach-teamcity committed:
orchestration: released CockroachDB version 24.1.2. Next version: 24.1.3

Release note: None
Epic: None
Release justification: non-production (release infra) change.

arulajmani committed:
kvserver: skip TestGossipHandlesReplacedNode under race

Closes #126815

Release note: None

navsetlur committed:
logical: Periodically send heartbeats to logical replication streams and retry ingestion on errors

There currently exists a heartbeat sender implementation in the logical job, but it isn't started
and doesn't handle errors currently. This change sets the heartbeat to be one of the goroutines
waited on in the Logical Replication Job and triggers retries on heartbeat failures.

Release note: none
Part of: https://cockroachlabs.atlassian.net/browse/CRDB-39983

cockroach-teamcity committed:
master: Update pkg/testutils/release/cockroach_releases.yaml

Update pkg/testutils/release/cockroach_releases.yaml with recent values.

Epic: None
Release note: None
Release justification: test-only updates

RaduBerinde committed:
go.mod: bump Pebble to 7920d968a3d5

Changes:

 * [`7920d968`](https://github.com/cockroachdb/pebble/commit/7920d968) *: add facility to print the iterator stack
 * [`d3cde23d`](https://github.com/cockroachdb/pebble/commit/d3cde23d) keyspan: reimplement straddle spans in LevelIter
 * [`e8fd26a0`](https://github.com/cockroachdb/pebble/commit/e8fd26a0) sstable: simplify monotonic bounds optimization condition
 * [`fad89cfb`](https://github.com/cockroachdb/pebble/commit/fad89cfb) rowblk: support synthetic prefix in FragmentIter
 * [`4981bd0e`](https://github.com/cockroachdb/pebble/commit/4981bd0e) colblk: add Bitmap, BitmapBuilder
 * [`a28d244d`](https://github.com/cockroachdb/pebble/commit/a28d244d) keyspan: remove err field from LevelIter
 * [`3ef2e5b1`](https://github.com/cockroachdb/pebble/commit/3ef2e5b1) rowblk: add a basic unit test for FragmentIter
 * [`144e5175`](https://github.com/cockroachdb/pebble/commit/144e5175) rowblk: reorganize tests
 * [`15ec0c9b`](https://github.com/cockroachdb/pebble/commit/15ec0c9b) sstable: add layoutWriter
 * [`89f4160b`](https://github.com/cockroachdb/pebble/commit/89f4160b) sstable: drop support for v1 range deletion blocks
 * [`870777c9`](https://github.com/cockroachdb/pebble/commit/870777c9) sstable: add metaIndexWriter
 * [`96660eb8`](https://github.com/cockroachdb/pebble/commit/96660eb8) sstable/rowblk: replace RawWriter with methods on rowblk.Writer
 * [`96eeb815`](https://github.com/cockroachdb/pebble/commit/96eeb815) vfs: fix memFile pos
 * [`18b77232`](https://github.com/cockroachdb/pebble/commit/18b77232) sstable: add FragmentIterTransforms

Release note: none.
Epic: none.

Commit:
drt: update drtprod start for drt-large

Previously, to start `drt-large` we had to run
```
./scripts/drtprod start drt-large --binary ./cockroach
--args=--log="file-defaults: {dir: 'logs',
max-group-size: 1GiB}" --store-count=16 --restart=false
```
This was inadequate for the uninitiated since it's unlikely
for them to know all these arguments. To address this, this
patch checks in the flags as default for `drt-large`

Epic: none
Fixes: #126807
Release note: None

yuzefovich committed:
colexec: fix IS NOT NULL filter for tuples with NULLs

This commit fixes `isTupleNullSelOp` (which handles IS NOT NULL filter
expression). Previously, we incorrectly short-circuited the filter
evaluation whenever the tuple is non-NULL - all such tuples were passed
by the filter even though only tuples that don't have any NULL elements
should pass. This is now fixed by bringing the logic inline with how
row-by-row engine does it as well as the `isTupleNullProjOp` does it.

This commit additionally extends an existing vectorized eval test (which
currently runs all projections) to also run filtering (i.e. "selection"
operators) on expressions of the boolean type. This would have caught
the bug.

Release note (bug fix): Previously, CockroachDB could incorrectly
evaluate `IS NOT NULL` filter if it was applied to non-NULL tuples that
had NULL elements (like `(1, NULL)` or `(NULL, NULL)`). The bug is
present since 20.2 version and is now fixed.

navsetlur committed:
logical: Refactor tests to use some common logic

There was setup between multiple tests that created a TestServer in the same way,
so it's been moved into its own function for reusability

Release note: none
Epic: none

pav-kv committed:
raft: update leader's Next when appending entries

This commit makes the leader update its Progress.Next when it appends
entries to its log. This makes the Progress.Next updates symmetric.

Previously, we only updated Next for followers, when sending a MsgApp to
them, but updated Match symmetrically when getting MsgAppResp (both from
the leader self-directed messages, and from followers). The leader is a
special case: it doesn't send MsgApps to itself. But there is an
implicit MsgApp that happens in appendEntry function. We update Next
accordingly.

On the leader, Next is always equal to log.lastIndex + 1. This is
consistent with the fact that no more MsgApps need to be sent to the
local Acceptor.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126903)
<!-- Reviewable:end -->

cty123 committed:
changefeedccl: add support for IAM role-based authentication to Kafka sink

This patch adds support to the go Kafka client(Sarama) to authenticate
with AWS IAM roles. AWS IAM role authentication is the recommended
way to connect with AWS MSK cluster as it's more secure than the
SASL username + password method.

AWS now provides a library in a repository under their official account
https://github.com/aws/aws-msk-iam-sasl-signer-go that is able to
connect to AWS MSK with IAM roles. This PR introduces the new
dependency on the library and provides option in Kafka DialConfig
to enable the IAM role authentcaition.

Resolves: #124431
Release note (enterprise change): Add support to the Kafka sink
to authenticate with MSK via AWS IAM roles

nvb committed:
kv: delete unused Store.spanConfigUpdateQueueRateLimiter

This has been unused since its introduction in 89f8aba8. For some
reason, it's now tripping up `TestLint/TestStaticCheck`.

Epic: None
Release note: None

pav-kv committed:
raft/testdata: log leader status after Ready

Epic: none
Release note: none

nvb committed:
raft: presize progress maps

Drive-by cleanup to pre-size the progress maps when constructing raft
Status and SparseStatus structs.

Epic: None
Release note: None

asg0451 committed:
changefeedccl: add roachtest testing msk serverless

Add a roachtest to test the kafka sink against msk serverless.

Resolves: #124431

Release note: None
